File: src/main/java/alfheim/client/render/tile/RenderTileAnimatedTorch.java
package alfheim.client.render.tile;

import static org.lwjgl.opengl.GL11.*;

import java.util.Random;

import alfheim.common.block.tile.TileAnimatedTorch;
import net.minecraft.client.Minecraft;
import net.minecraft.client.renderer.Tessellator;
import net.minecraft.client.renderer.entity.RenderItem;
import net.minecraft.client.renderer.entity.RenderManager;
import net.minecraft.client.renderer.tileentity.TileEntitySpecialRenderer;
import net.minecraft.entity.item.EntityItem;
import net.minecraft.init.Blocks;
import net.minecraft.item.ItemStack;
import net.minecraft.tileentity.TileEntity;
import vazkii.botania.client.core.handler.ClientTickHandler;

public class RenderTileAnimatedTorch extends TileEntitySpecialRenderer {

	@Override
	public void renderTileEntityAt(TileEntity te, double x, double y, double z, float partialTicks) {
		render((TileAnimatedTorch) te, x, y, z, partialTicks);
	}
	
	Random rand = new Random();
	
	public void render(TileAnimatedTorch tile, double x, double y, double z, float partialTicks) {
		Minecraft mc = Minecraft.getMinecraft();
		glPushMatrix();
		glTranslated(x, y, z);

		boolean hasWorld = tile != null && tile.getWorldObj() != null;
		int wtime = !hasWorld ? 0 : ClientTickHandler.ticksInGame;
		
		if(wtime != 0) {
			rand.setSeed(tile.xCoord ^ tile.yCoord ^ tile.zCoord);
			wtime += rand.nextInt(360);
		}
			
		float time = wtime == 0 ? 0 : wtime + partialTicks;
		double xt = 0.5 + Math.cos(time * 0.05) * 0.025;
		double yt = 0.1 + (Math.sin(time * 0.04) + 1) * 0.05;
		double zt = 0.5 + Math.sin(time * 0.05) * 0.025;
		glTranslated(xt, yt, zt);

		glScaled(2, 2, 2);
		glRotated(90, 1, 0, 0);
		float rotation = (float) tile.rotation;
		if(tile.rotating)
			rotation += tile.anglePerTick * partialTicks;

		glRotated(rotation, 0F, 0F, 1F);
		glTranslated(0, 0.15, 0);
		
		{
			glDisable(GL_CULL_FACE);
			EntityItem entityitem = new EntityItem(hasWorld ? tile.getWorldObj() : Minecraft.getMinecraft().theWorld, 0.0, 0.0, 0.0, new ItemStack(Blocks.redstone_torch));
			entityitem.hoverStart = 0.0F;
			glPushMatrix();
			Tessellator.instance.setBrightness(tile.getBlockType().getMixedBrightnessForBlock(tile.getWorldObj(), tile.xCoord, tile.yCoord, tile.zCoord));
			RenderItem.renderInFrame = true;
			RenderManager.instance.renderEntityWithPosYaw(entityitem, 0.0, -0.2501, 0.0, 0.0F, 0.0F);
			RenderItem.renderInFrame = false;
			glEnable(GL_CULL_FACE);
			glPopMatrix();
		}
		
		glPopMatrix();
	}
}

File: src/main/java/alfheim/client/render/tile/RenderTileAnimatedTorch.kt
package alfheim.client.render.tile

import alexsocol.asjlib.*
import alfheim.common.block.tile.TileAnimatedTorch
import net.minecraft.client.renderer.Tessellator
import net.minecraft.client.renderer.entity.*
import net.minecraft.client.renderer.tileentity.TileEntitySpecialRenderer
import net.minecraft.entity.item.EntityItem
import net.minecraft.init.Blocks
import net.minecraft.item.ItemStack
import net.minecraft.tileentity.TileEntity
import org.lwjgl.opengl.GL11.*
import vazkii.botania.client.core.handler.ClientTickHandler
import java.util.*
import kotlin.math.*

object RenderTileAnimatedTorch: TileEntitySpecialRenderer() {
	
	val rand = Random()
	
	override fun renderTileEntityAt(tile: TileEntity, x: Double, y: Double, z: Double, partialTicks: Float) {
		if (tile !is TileAnimatedTorch) return
		
		glPushMatrix()
		glTranslated(x, y, z)
		
		val hasWorld = tile.worldObj != null
		var wtime = if (!hasWorld) 0 else ClientTickHandler.ticksInGame
		
		if (wtime != 0) {
			rand.setSeed((tile.xCoord xor tile.yCoord xor tile.zCoord).toLong())
			wtime += rand.nextInt(360)
		}
		
		val time = if (wtime == 0) 0f else wtime + partialTicks
		val xt = 0.5 + cos(time * 0.05) * 0.025
		val yt = 0.1 + (sin(time * 0.04) + 1) * 0.05
		val zt = 0.5 + sin(time * 0.05) * 0.025
		glTranslated(xt, yt, zt)
		
		glScaled(2.0)
		glRotated(90.0, 1.0, 0.0, 0.0)
		
		var rotation = tile.rotation.F
		if (tile.rotating)
			rotation += (tile.anglePerTick * partialTicks).F
		
		glRotated(rotation.D, 0.0, 0.0, 1.0)
		glTranslated(0.0, 0.15, 0.0)
		
		run {
			glDisable(GL_CULL_FACE)
			val entityitem = EntityItem(if (hasWorld) tile.worldObj else mc.theWorld, 0.0, 0.0, 0.0, ItemStack(Blocks.redstone_torch))
			entityitem.hoverStart = 0f
			glPushMatrix()
			Tessellator.instance.setBrightness(tile.getBlockType().getMixedBrightnessForBlock(tile.worldObj, tile.xCoord, tile.yCoord, tile.zCoord))
			RenderItem.renderInFrame = true
			RenderManager.instance.renderEntityWithPosYaw(entityitem, 0.0, -0.2501, 0.0, 0f, 0f)
			RenderItem.renderInFrame = false
			glEnable(GL_CULL_FACE)
			glPopMatrix()
		}
		
		glPopMatrix()
	}
}
